### ISSUE TITLE
perf(db): SQLite en 710 MB y ciclos del live job en 9-12 min contra schedulers de 15

### ISSUE THREAD

**Issue by LorenzoGalaverna** [santiago6124/OnTimeAI-Backend#4]
## Resumen

La base SQLite alcanzó **710 MB** y el ciclo del live job tarda entre **9 y 12 minutos**, contra schedulers que disparan cada **15**. El margen se está agotando.

## Evidencia

`GET /admin/db-stats` en producción (2026-09-04):

```json
{"db_size_mb": 710.21,
 "table_counts": {"predictions": 192578, "actuals": 397604,
                  "flights": 400250, "prediction_shap": 879720,
                  "weather_obs": 102995, "aircraft_position": 76093,
                  "runs": 2294},
 "prediction_dates": {"first": "2026-08-06", "last": "2026-09-04"}}
```

Duración de las últimas 6 ejecuciones de `ontimeai-live-pull`:

```
557s   712s   665s   639s   355s   414s
```

## Por qué pasa

El job baja la DB entera de GCS, la modifica y la vuelve a subir en cada ciclo. Con 710 MB, la transferencia por sí sola domina el tiempo de ejecución. Ese costo crece linealmente con el tamaño del archivo.

## Proyección

29 días de datos → 710 MB, o sea **~24 MB/día**:

| Horizonte | Tamaño estimado |
|---|---|
| +30 días | ~1,4 GB |
| +60 días | ~2,2 GB |
| +90 días | ~2,9 GB |

El `--task-timeout` actual es de 1800s. Antes de llegar ahí, los ciclos van a solaparse entre sí: dos jobs escribiendo la misma DB y pisándose las subidas a GCS.

`prediction_shap` (879.720 filas) y `flights` (400.250) son las tablas dominantes.

## Direcciones posibles

- **Retención**: purgar `prediction_shap` y `aircraft_position` más allá de N días; son las que más crecen y las de menor valor histórico
- **Separar el histórico**: mover lo que no se consulta en vivo a un archivo aparte en GCS, o a BigQuery
- **Evitar el round-trip completo**: no bajar/subir el archivo entero en cada ciclo
- **VACUUM periódico**: recuperar espacio de filas borradas

## Criterios de aceptación

- [ ] El ciclo del live job vuelve a estar consistentemente por debajo de los 5 minutos
- [ ] El tamaño de la DB queda acotado por una política de retención explícita
- [ ] Existe alerta o umbral visible cuando el tamaño o la duración se acercan al límite
- [ ] La purga no rompe el cálculo de AUC live (que usa los últimos 7 días de actuals)

**Comment by LorenzoGalaverna** [comment]
Cierro con el balance medido. La investigación completa quedó en `docs/CICLO_Y_FRECUENCIA.md`.

## Criterios

- [x] **El tamaño de la DB queda acotado por una política de retención explícita** — documentada en `docs/RETENCION.md` (#12), con agregados diarios que sobreviven a la purga.
- [x] **Existe alerta o umbral visible cuando el tamaño o la duración se acercan al límite** — el vigía alerta por tamaño (>800 MB) y por duración (mediana >10 min, dos tercios de la ventana del scheduler). Mide el ciclo completo: la primera versión medía solo el pipeline y no se habría disparado nunca.
- [x] **La purga no rompe el cálculo de AUC live** — verificado.
- [ ] **El ciclo consistentemente por debajo de los 5 minutos** — hoy 5,1 a 8,9.

## De paso, lo que sí bajó

```
base    722 → 468 MB    (barrido de 782k filas huérfanas, #46)
imagen  353 → 247 MB    (#54 y #55)
```

## Por qué el cuarto criterio no se alcanza

El ciclo, medido por fase desde dentro del job:

```
importar dependencias :    3,0 s
descarga de GCS       :    7,8 s
pipeline              :  176,2 s   ← clima 64 s + features/inferencia 91 s
agregados diarios     :    0,7 s
purga                 :    1,0 s
subida a GCS          :    8,9 s
                        ─────────
ciclo medido          :    3,1 min
ejecución de Cloud Run: 5,1 - 8,9 min
```

Cinco hipótesis descartadas, cada una con el dato que la descartó:

| hipótesis | medición |
|---|---|
| la transferencia de la base domina | **17 s** ida y vuelta |
| los agregados diarios son caros | 0,7 s |
| la purga es cara | 1,0 s |
| importar pandas/lightgbm/sklearn es caro | 3,0 s |
| el job reintenta tareas | `retriedCount` vacío |

La premisa del issue — *"con 710 MB, la transferencia por sí sola domina el tiempo de ejecución"* — es **empíricamente falsa**.

Lo que queda es el arranque en frío, 2 a 6 minutos, y **no se pudo reducir**:

```
imagen 353 MB  → 120,4 s
imagen 320 MB  → 129,4 s
imagen 247 MB  → 323,6  291,6  239,0  292,7  125,1  153,6 s
```

Una imagen recién construida arranca dos o tres veces más lento durante las primeras corridas. Las últimas dos de la imagen chica caen en el mismo rango que las grandes: **con caché caliente las tres tardan lo mismo**. Recortar el 30% no cambió el arranque.

## Propuesta para el criterio

Lo que ese criterio protegía era el **solapamiento** — que dos jobs escriban la misma base y se pisen las subidas. Eso está cubierto por otras vías:

- el scheduler dispara cada 15 min y los ciclos duran 5-9: hay margen
- el vigía alerta si la mediana pasa los 10
- cada job toca la base 2-3 min de los 7 que dura, así que solaparse en reloj no implica chocar
- **cero conflictos CAS** medidos en 4 horas

Propongo reemplazarlo por: *"el ciclo no se acerca a la ventana del scheduler, y hay alerta si lo hace"*. Bajarlo de verdad exigiría optimizar el pipeline —código que funciona— o mudarse de Cloud Run Jobs, y ninguna se justifica por el riesgo que queda.

Si preferís que quede abierto hasta discutirlo, reabrilo sin problema.

## Sobre correr más seguido

**No se puede sin colisiones.** La hora está completa:

```
live-pull     :00-:07   :30-:38
live-pull-2   :15-:21   :45-:52
harvester     :08-:14   :23-:28   :38-:46   :53-:58
```

El margen más ajustado es de un minuto. Subir la frecuencia exige primero acortar el ciclo, en ese orden.

## Un hallazgo secundario, anotado en el doc

Las primeras corridas después de **cada despliegue** arrancan dos o tres veces más lento. En un día con varios despliegues seguidos eso infla los ciclos y parece degradación del pipeline cuando es caché fría. Antes de investigar un ciclo lento, conviene mirar si hubo un despliegue reciente.
